File: android/app/src/main/java/com/noop/push/PushBinaryCodec.kt
package com.noop.push

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest

/** Platform-neutral packed payload for `bin_gzip_noop_push_v1` / `protobuf_zstd_noop_push_v1`. */
object PushBinaryCodec {
    val MAGIC = "NPB1".toByteArray(Charsets.US_ASCII)
    const val FORMAT_VERSION: Byte = 1
    const val IMU_COLUMNS_PER_RECORD = 600
    const val IMU_RECORD_PAYLOAD_BYTES = IMU_COLUMNS_PER_RECORD * 2

    enum class Kind(val raw: Byte) {
        PPG_WAVEFORM_SAMPLE(1),
        V18_AUX_SAMPLE(2),
        RAW_BATCH(3),
        RAW_IMU_SESSION(4),
    }

    fun kind(table: PushBinaryTable): Kind = when (table) {
        PushBinaryTable.PPG_WAVEFORM_SAMPLE -> Kind.PPG_WAVEFORM_SAMPLE
        PushBinaryTable.V18_AUX_SAMPLE -> Kind.V18_AUX_SAMPLE
        PushBinaryTable.RAW_BATCH -> Kind.RAW_BATCH
        PushBinaryTable.RAW_IMU_SESSION -> Kind.RAW_IMU_SESSION
    }

    fun pack(table: PushBinaryTable, rows: List<PushBinaryRow>, ppgIdentityV2: Boolean = false, auxIdentityV2: Boolean = false): ByteArray {
        if (rows.isEmpty()) throw PushProtocolException("binary object must contain a row")
        return when (table) {
            PushBinaryTable.PPG_WAVEFORM_SAMPLE -> {
                val records = rows.map { row ->
                    when (row) {
                        is PushBinaryRow.PpgWaveform -> row.record
                        else -> throw PushProtocolException("binary row kind mismatch")
                    }
                }
                packPpgRecords(records, ppgIdentityV2)
            }
            PushBinaryTable.V18_AUX_SAMPLE -> {
                val records = rows.map { row ->
                    when (row) {
                        is PushBinaryRow.V18Aux -> row.record
                        else -> throw PushProtocolException("binary row kind mismatch")
                    }
                }
                packV18Records(records, auxIdentityV2)
            }
            PushBinaryTable.RAW_BATCH -> {
                if (rows.size != 1) throw PushProtocolException("rawBatch upload must contain exactly one batch row")
                when (val row = rows.single()) {
                    is PushBinaryRow.RawBatch -> packRawBatch(row.record)
                    else -> throw PushProtocolException("rawBatch upload must contain exactly one batch row")
                }
            }
            PushBinaryTable.RAW_IMU_SESSION -> {
                val records = rows.map { row ->
                    when (row) {
                        is PushBinaryRow.RawImuSession -> row.record
                        else -> throw PushProtocolException("binary row kind mismatch")
                    }
                }
                packImuRecords(records)
            }
        }
    }

    fun packedHeaderSize(table: PushBinaryTable): Int = when (table) {
        PushBinaryTable.RAW_BATCH -> 6
        PushBinaryTable.PPG_WAVEFORM_SAMPLE, PushBinaryTable.V18_AUX_SAMPLE, PushBinaryTable.RAW_IMU_SESSION -> 10
    }

    fun packedRowSize(row: PushBinaryRow, ppgIdentityV2: Boolean = false, auxIdentityV2: Boolean = false): Int = when (row) {
        is PushBinaryRow.PpgWaveform -> {
            val record = row.record
            8 + 8 + (if (ppgIdentityV2) (if (record.recordIndex == null) 1 else 9) else 0) + 1 + (if (record.burstIndex != null) 4 else 0) + 4 + record.samples.size
        }
        is PushBinaryRow.V18Aux -> 8 + 8 + (if (auxIdentityV2) (if (row.record.recordIndex == null) 1 else 9) else 0) + 4 + row.record.fields.size
        is PushBinaryRow.RawImuSession -> {
            if (row.record.columns.size != IMU_RECORD_PAYLOAD_BYTES) {
                throw PushProtocolException("rawImuSession record must carry $IMU_COLUMNS_PER_RECORD i16 columns")
            }
            8 + 8 + 4 + row.record.columns.size
        }
        is PushBinaryRow.RawBatch -> {
            val record = row.record
            2 + record.batchId.toByteArray(Charsets.UTF_8).size + 8 * 5 + 4 + 4 + 4 + record.framesBlob.size
        }
    }

    private fun packPpgRecords(records: List<PushPpgWaveformRecord>, identityV2: Boolean): ByteArray {
        if (!identityV2 && records.any { it.recordIndex != null }) throw PushProtocolException("PPG identity requires negotiated protocol 1.3")
        val out = ByteArrayOutputStream(records.size * 32 + 8)
        out.write(MAGIC); out.write(if (identityV2) 2 else 1); out.write(Kind.PPG_WAVEFORM_SAMPLE.raw.toInt())
        appendU32(records.size, out)
        for (record in records) {
            appendI64(record.rowId, out)
            appendI64(record.ts, out)
            if (identityV2) {
                val index = record.recordIndex
                if (index == null) out.write(0) else {
                    if (index !in 0..4294967295L) throw PushProtocolException("invalid PPG record index")
                    out.write(1); appendI64(index, out)
                }
            }
            if (record.burstIndex != null) {
                out.write(1)
                appendI32(record.burstIndex, out)
            } else {
                out.write(0)
            }
            appendBlob(record.samples, out)
        }
        return out.toByteArray()
    }

    private fun packV18Records(records: List<PushV18AuxRecord>, identityV2: Boolean): ByteArray {
        if (!identityV2 && records.any { it.recordIndex != null || com.noop.data.V18AuxIdentity.complete(it.fields)?.recordIndex != null })
            throw PushProtocolException("Auxiliary identity requires negotiated protocol 1.4")
        val out = ByteArrayOutputStream(records.size * 32 + 8)
        out.write(MAGIC); out.write(if (identityV2) 2 else 1); out.write(Kind.V18_AUX_SAMPLE.raw.toInt())
        appendU32(records.size, out)
        for (record in records) {
            appendI64(record.rowId, out)
            appendI64(record.ts, out)
            if (identityV2) {
                val complete = com.noop.data.V18AuxIdentity.complete(record.fields)
                    ?: throw PushProtocolException("Auxiliary fields require raw-archive validation")
                if (complete.recordIndex != record.recordIndex) throw PushProtocolException("Auxiliary envelope identity mismatch")
                if (record.recordIndex == null) out.write(0) else {
                    if (record.recordIndex !in 0..4294967295L) throw PushProtocolException("Invalid auxiliary index")
                    out.write(1); appendI64(record.recordIndex, out)
                }
            }
            appendBlob(record.fields, out)
        }
        return out.toByteArray()
    }

    private fun packRawBatch(record: PushRawBatchRecord): ByteArray {
        val out = ByteArrayOutputStream(record.framesBlob.size + 64)
        header(Kind.RAW_BATCH, out)
        appendUtf8(record.batchId, out)
        appendI64(record.capturedAt, out)
        appendI64(record.deviceClockRef, out)
        appendI64(record.wallClockRef, out)
        appendI64(record.startTs, out)
        appendI64(record.endTs, out)
        appendI32(record.frameCount, out)
        appendI32(record.byteSize, out)
        appendBlob(record.framesBlob, out)
        return out.toByteArray()
    }

    private fun packImuRecords(records: List<PushRawImuRecord>): ByteArray {
        val out = ByteArrayOutputStream(records.size * (IMU_RECORD_PAYLOAD_BYTES + 24) + 8)
        header(Kind.RAW_IMU_SESSION, out)
        appendU32(records.size, out)
        for (record in records) {
            if (record.columns.size != IMU_RECORD_PAYLOAD_BYTES) {
                throw PushProtocolException("rawImuSession record must carry $IMU_COLUMNS_PER_RECORD i16 columns")
            }
            appendI64(record.rowId, out)
            appendI64(record.ts, out)
            appendBlob(record.columns, out)
        }
        return out.toByteArray()
    }

    private fun header(kind: Kind, out: ByteArrayOutputStream) {
        out.write(MAGIC)
        out.write(FORMAT_VERSION.toInt())
        out.write(kind.raw.toInt())
    }

    private fun appendBlob(blob: ByteArray, out: ByteArrayOutputStream) {
        if (blob.size > PushProtocol.MAX_BODY_BYTES) {
            throw PushProtocolException("binary blob exceeds decoded limit")
        }
        appendU32(blob.size, out)
        out.write(blob)
    }

    private fun appendUtf8(value: String, out: ByteArrayOutputStream) {
        val bytes = value.toByteArray(Charsets.UTF_8)
        if (bytes.size > UShort.MAX_VALUE.toInt()) {
            throw PushProtocolException("binary string exceeds limit")
        }
        out.write(
            ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN).putShort(bytes.size.toShort()).array(),
        )
        out.write(bytes)
    }

    private fun appendI32(value: Int, out: ByteArrayOutputStream) {
        out.write(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(value).array())
    }

    private fun appendU32(value: Int, out: ByteArrayOutputStream) = appendI32(value, out)

    private fun appendI64(value: Long, out: ByteArrayOutputStream) {
        out.write(ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN).putLong(value).array())
    }

    fun sha256Hex(bytes: ByteArray): String = MessageDigest.getInstance("SHA-256")
        .digest(bytes).joinToString("") { "%02x".format(it.toInt() and 0xff) }
}

File: Packages/NoopPush/Sources/NoopPush/PushBinaryCodec.swift
import Foundation
import CryptoKit

/// Platform-neutral packed payload for `bin_gzip_noop_push_v1` / `protobuf_zstd_noop_push_v1`.
public enum PushBinaryCodec {
    public static let magic = Data("NPB1".utf8)
    public static let formatVersion: UInt8 = 1

    public enum Kind: UInt8 {
        case ppgWaveformSample = 1
        case v18AuxSample = 2
        case rawBatch = 3
        case rawImuSession = 4
    }

    /// 100 samples/second × 6 axes, axis-major i16 LE — the record layout ImuSessionFileStore writes.
    public static let imuColumnsPerRecord = 600
    public static let imuRecordPayloadBytes = imuColumnsPerRecord * 2

    public static func kind(for table: PushBinaryTable) -> Kind {
        switch table {
        case .ppgWaveformSample: return .ppgWaveformSample
        case .v18AuxSample: return .v18AuxSample
        case .rawBatch: return .rawBatch
        case .rawImuSession: return .rawImuSession
        }
    }

    public static func pack(table: PushBinaryTable, rows: [PushBinaryRow], ppgIdentityV2: Bool = false,
                            v18IdentityV2: Bool = false) throws -> Data {
        guard !rows.isEmpty else { throw PushProtocolException("binary object must contain a row") }
        switch table {
        case .ppgWaveformSample:
            let records = try rows.map { row -> PushPpgWaveformRecord in
                guard case .ppgWaveform(let record) = row else {
                    throw PushProtocolException("binary row kind mismatch")
                }
                return record
            }
            return try packPpgRecords(records, identityV2: ppgIdentityV2)
        case .v18AuxSample:
            let records = try rows.map { row -> PushV18AuxRecord in
                guard case .v18Aux(let record) = row else {
                    throw PushProtocolException("binary row kind mismatch")
                }
                return record
            }
            return try packV18Records(records, identityV2: v18IdentityV2)
        case .rawBatch:
            guard rows.count == 1, case .rawBatch(let record) = rows[0] else {
                throw PushProtocolException("rawBatch upload must contain exactly one batch row")
            }
            return try packRawBatch(record)
        case .rawImuSession:
            let records = try rows.map { row -> PushRawImuRecord in
                guard case .rawImuSession(let record) = row else {
                    throw PushProtocolException("binary row kind mismatch")
                }
                return record
            }
            return try packImuRecords(records)
        }
    }

    /// Exact packed size of the header for a table, without packing. Counted kinds carry a u32
    /// record count after the 6-byte magic/version/kind header; rawBatch does not.
    public static func packedHeaderSize(for table: PushBinaryTable) -> Int {
        switch table {
        case .rawBatch: return 6
        case .ppgWaveformSample, .v18AuxSample, .rawImuSession: return 10
        }
    }

    /// Exact packed size of one row, without packing. Lets batch selection run O(n) instead of
    /// re-packing the growing candidate set per row.
    public static func packedRowSize(_ row: PushBinaryRow, ppgIdentityV2: Bool = false,
                                      v18IdentityV2: Bool = false) throws -> Int {
        switch row {
        case .ppgWaveform(let record):
            return 8 + 8 + 1 + (record.burstIndex != nil ? 4 : 0) + 4 + record.samples.count
                + (ppgIdentityV2 ? (record.recordIndex == nil ? 1 : 9) : 0)
        case .v18Aux(let record):
            return 8 + 8 + 4 + record.fields.count + (v18IdentityV2 ? (record.recordIndex == nil ? 1 : 9) : 0)
        case .rawImuSession(let record):
            guard record.columns.count == imuRecordPayloadBytes else {
                throw PushProtocolException("rawImuSession record must carry \(imuColumnsPerRecord) i16 columns")
            }
            return 8 + 8 + 4 + record.columns.count
        case .rawBatch(let record):
            return 2 + record.batchId.utf8.count + 8 * 5 + 4 + 4 + 4 + record.framesBlob.count
        }
    }

    private static func packPpgRecords(_ records: [PushPpgWaveformRecord], identityV2: Bool) throws -> Data {
        guard identityV2 || records.allSatisfy({ $0.recordIndex == nil }) else {
            throw PushProtocolException("PPG record identity requires negotiated protocol 1.3")
        }
        var out = header(kind: .ppgWaveformSample)
        if identityV2 { out[4] = 2 }
        appendU32(Int32(records.count), to: &out)
        for record in records {
            appendI64(record.rowId, to: &out)
            appendI64(record.ts, to: &out)
            if identityV2 {
                if let index = record.recordIndex {
                    guard index >= 0 else { throw PushProtocolException("invalid PPG record index") }
                    out.append(1)
                    appendI64(index, to: &out)
                } else { out.append(0) }
            }
            if let burstIndex = record.burstIndex {
                out.append(1)
                appendI32(burstIndex, to: &out)
            } else {
                out.append(0)
            }
            try appendBlob(record.samples, to: &out)
        }
        return out
    }

    private static func packV18Records(_ records: [PushV18AuxRecord], identityV2: Bool) throws -> Data {
        guard identityV2 || records.allSatisfy({ $0.recordIndex == nil }) else {
            throw PushProtocolException("auxiliary record identity requires negotiated protocol 1.4")
        }
        var out = header(kind: .v18AuxSample)
        if identityV2 { out[4] = 2 }
        appendU32(Int32(records.count), to: &out)
        for record in records {
            appendI64(record.rowId, to: &out)
            appendI64(record.ts, to: &out)
            if identityV2 {
                if let index = record.recordIndex {
                    guard (0...Int64(UInt32.max)).contains(index) else { throw PushProtocolException("invalid auxiliary record index") }
                    out.append(1); appendI64(index, to: &out)
                } else { out.append(0) }
            }
            try appendBlob(record.fields, to: &out)
        }
        return out
    }

    private static func packRawBatch(_ record: PushRawBatchRecord) throws -> Data {
        var out = header(kind: .rawBatch)
        try appendUtf8(record.batchId, to: &out)
        appendI64(record.capturedAt, to: &out)
        appendI64(record.deviceClockRef, to: &out)
        appendI64(record.wallClockRef, to: &out)
        appendI64(record.startTs, to: &out)
        appendI64(record.endTs, to: &out)
        appendI32(record.frameCount, to: &out)
        appendI32(record.byteSize, to: &out)
        try appendBlob(record.framesBlob, to: &out)
        return out
    }

    private static func packImuRecords(_ records: [PushRawImuRecord]) throws -> Data {
        var out = header(kind: .rawImuSession)
        appendU32(Int32(records.count), to: &out)
        for record in records {
            guard record.columns.count == imuRecordPayloadBytes else {
                throw PushProtocolException("rawImuSession record must carry \(imuColumnsPerRecord) i16 columns")
            }
            appendI64(record.rowId, to: &out)
            appendI64(record.ts, to: &out)
            try appendBlob(record.columns, to: &out)
        }
        return out
    }

    private static func header(kind: Kind) -> Data {
        var out = Data(capacity: 6)
        out.append(magic)
        out.append(formatVersion)
        out.append(kind.rawValue)
        return out
    }

    private static func appendBlob(_ blob: Data, to out: inout Data) throws {
        guard blob.count <= PushProtocolLimits.maxBodyBytes else {
            throw PushProtocolException("binary blob exceeds decoded limit")
        }
        appendU32(Int32(blob.count), to: &out)
        out.append(blob)
    }

    private static func appendUtf8(_ value: String, to out: inout Data) throws {
        let bytes = Data(value.utf8)
        guard bytes.count <= Int(UInt16.max) else {
            throw PushProtocolException("binary string exceeds limit")
        }
        var length = UInt16(bytes.count).littleEndian
        withUnsafeBytes(of: &length) { out.append(contentsOf: $0) }
        out.append(bytes)
    }

    private static func appendI32(_ value: Int32, to out: inout Data) {
        var le = value.littleEndian
        withUnsafeBytes(of: &le) { out.append(contentsOf: $0) }
    }

    private static func appendU32(_ value: Int32, to out: inout Data) {
        appendI32(value, to: &out)
    }

    private static func appendI64(_ value: Int64, to out: inout Data) {
        var le = value.littleEndian
        withUnsafeBytes(of: &le) { out.append(contentsOf: $0) }
    }

    public static func sha256Hex(_ bytes: Data) -> String {
        SHA256.hash(data: bytes).map { String(format: "%02x", $0) }.joined()
    }
}
